pandas/core/indexes/datetimelike.py

```diff
@@ -205,6 +205,43 @@ def floor(self, freq):
     def ceil(self, freq):
         return self._round(freq, np.ceil)
 
+    @classmethod
+    def _validate_frequency(cls, index, freq, **kwargs):
+        """
+        Validate that a frequency is compatible with the values of a given
+        DatetimeIndex or TimedeltaIndex
+
+        Parameters
+        ----------
+        index : DatetimeIndex or TimedeltaIndex
+            The index on which to determine if the given frequency is valid
+        freq : DateOffset
+            The frequency to validate
+        """
+        inferred = index.inferred_freq
+        if index.empty or inferred == freq.freqstr:
+            return None
+
+        on_freq = cls._generate(
+            index[0], None, len(index), None, freq, **kwargs)
```

Comment:
is generating the index needed to know if it a correct frequence? Checking the inferred frequency is not enough?

Reply:
OK, see it was also done like that below (still wondering though)

Reply:
I think it's necessary, at least to a certain extent.  There are cases where multiple frequencies are valid for a given set of dates, so you can have a valid frequency that's not the inferred frequency, e.g. `['2018-01-01', '2018-01-02', '2018-01-03']` could be calendar day or business day.  Users can also define custom frequencies that would need to be validated but would not be the inferred frequency.

There are cases where you don't need to generate the entire index to reject invalid frequencies, e.g. if the second generated value doesn't match.  But I don't immediately see how to get around generating the entire index to determine if a frequency is valid.  I suppose you could maybe optimize memory usage with huge indexes by doing some type of elementwise validation, only keeping the current generated element in memory.

Reply:
Ah, yes, I see, that's for sure a valid case.

For me, improving the performance of the validation is not very high priority.